runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/BatchFlinkExecutableStageContext.java

```diff
@@ -0,0 +1,107 @@
+import org.apache.beam.runners.flink.ArtifactSourcePool;
+import org.apache.beam.runners.fnexecution.control.DockerJobBundleFactory;
+import org.apache.beam.runners.fnexecution.control.JobBundleFactory;
+import org.apache.beam.runners.fnexecution.control.StageBundleFactory;
+import org.apache.beam.runners.fnexecution.provisioning.JobInfo;
+import org.apache.beam.runners.fnexecution.state.StateRequestHandler;
+import org.apache.flink.api.common.functions.RuntimeContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Implementation of a {@link FlinkExecutableStageContext} for batch jobs. */
+class BatchFlinkExecutableStageContext implements FlinkExecutableStageContext {
+  private static final Logger LOG = LoggerFactory.getLogger(BatchFlinkExecutableStageContext.class);
+
+  private final JobBundleFactory jobBundleFactory;
+  private final ArtifactSourcePool artifactSourcePool;
+
+  private static BatchFlinkExecutableStageContext create(JobInfo jobInfo) throws Exception {
+    ArtifactSourcePool artifactSourcePool = ArtifactSourcePool.create();
+    JobBundleFactory jobBundleFactory = DockerJobBundleFactory.create(jobInfo, artifactSourcePool);
+    return new BatchFlinkExecutableStageContext(jobBundleFactory, artifactSourcePool);
+  }
+
+  private BatchFlinkExecutableStageContext(
+      JobBundleFactory jobBundleFactory, ArtifactSourcePool artifactSourcePool) {
+    this.jobBundleFactory = jobBundleFactory;
+    this.artifactSourcePool = artifactSourcePool;
+  }
+
+  @Override
+  public <InputT> StageBundleFactory getStageBundleFactory(ExecutableStage executableStage) {
+    return jobBundleFactory.<InputT>forStage(executableStage);
+  }
+
+  @Override
+  public StateRequestHandler getStateRequestHandler(
+      ExecutableStage executableStage, RuntimeContext runtimeContext) {
+    return FlinkBatchStateRequestHandler.forStage(executableStage, runtimeContext);
+  }
+
+  @Override
+  public ArtifactSourcePool getArtifactSourcePool() {
+    return artifactSourcePool;
+  }
+
+  private void cleanUp() throws Exception {
+    jobBundleFactory.close();
+  }
+
+  enum BatchFactory implements Factory {
+    INSTANCE;
+
+    private final LoadingCache<JobInfo, BatchFlinkExecutableStageContext> cachedContexts;
+
+    BatchFactory() {
+      cachedContexts =
+          CacheBuilder.newBuilder()
+              .weakValues()
+              .removalListener(
+                  (RemovalNotification<JobInfo, BatchFlinkExecutableStageContext> removal) -> {
+                    try {
+                      removal.getValue().cleanUp();
+                    } catch (Exception e) {
+                      LOG.warn(
+                          "Error cleaning up bundle factory for job " + removal.getKey().jobId(),
+                          e);
+                    }
+                  })
+              .build(
+                  new CacheLoader<JobInfo, BatchFlinkExecutableStageContext>() {
```

Comment:
Can this be a lambda too?

Reply:
It could in theory (since it only has a single abstract method), but `CacheLoader` is not explicitly tagged as a `@FunctionalInterface`, which means that using a lambda may break in future Guava releases.